sdk/python/feast/registry.py

```diff
@@ -0,0 +1,123 @@
+                if existing_entity_proto.spec.name == entity_proto.spec.name:
+                    registry_proto.entities[idx] = entity_proto
+                    return registry_proto
+            registry_proto.entities.append(entity_proto)
+            return registry_proto
+        self._update_registry(updater)
+        return
+
+    def list_entities(self):
+        registry_proto = self._get_registry()
+        entities = []
+        for entity_proto in registry_proto.entities:
+            entities.append(Entity.from_proto(entity_proto))
+        return entities
+
+    def get_entity(self, name: str):
+        registry_proto = self._get_registry()
+        for entity_proto in registry.entities:
+            if entity_proto.spec.name == name:
+                return Entity.from_proto(entity_proto)
+        raise Exception(f"Entity {name} does not exist")
+
+    def apply_feature_table(self, feature_table: FeatureTable):
+        feature_table.is_valid()
+        feature_table_proto = feature_table.to_proto()
+        def updater(registry_proto: RegistryProto):
+            for idx, existing_feature_table_proto in enumerate(registry_proto.feature_tables):
+                if existing_feature_table_proto.spec.name == feature_table_proto.spec.name:
+                    registry_proto.feature_tables[idx] = feature_table_proto
+                    return registry_proto
+            registry_proto.feature_tables.append(feature_table_proto)
+            return registry_proto
+        self._update_registry(updater)
+        return
+
+    def list_feature_tables(self):
+        registry_proto = self._get_registry()
+        feature_tables = []
+        for feature_table_proto in registry_proto.feature_tables:
+            feature_tables.append(FeatureTable.from_proto(feature_table_proto))
+        return feature_tables
+
+    def get_feature_table(self, name: str):
+        registry_proto = self._get_registry()
+        for feature_table_proto in registry_proto.feature_tables:
+            if feature_table_proto.spec.name == name:
+                return FeatureTable.from_proto(feature_table_proto)
+        raise Exception(f"Feature table {name} does not exist")
+
+    def delete_feature_table(self, name: str):
+        def updater(registry_proto: RegistryProto):
+            for idx, existing_feature_table_proto in enumerate(registry_proto.feature_tables):
+                if existing_feature_table_proto.spec.name == feature_table_proto.spec.name:
+                    del registry_proto.feature_tables[idx]
+                    return registry_proto
+            raise Exception(f"Feature table {name} does not exist")
+        self._update_registry(updater)
+        return
+
+    @abstractmethod
+    def _get_registry(self):
+        pass
+
+    @abstractmethod
+    def _update_registry(self, updater: Callable[[RegistryProto], RegistryProto]):
+        pass
+
+
+class LocalRegistry(Registry):
```

Comment:
I'd avoid having class hierarchies (except for pure interfaces) as a general rule for better readability. They tend to make things unnecessarily hard to follow when the control flow jumps between the descendant class and the parent. Right now it is fairly simple but as this code will evolve it will likely make things complicated.

One idea would be to separate this into a Registry class (that wouldn't inherit from anything) and maybe RegistryStore(pure ABC), and then LocalRegistryStore would implement get_registry and update_registry. And Registry would use RegistryStore.

Reply:
@jklegar can we start with this proposal ^?